engine/resources/docker.go

```diff
@@ -0,0 +1,414 @@
+// Mgmt
+// Copyright (C) 2013-2018+ James Shubin and the project contributors
+// Written by James Shubin <james@shubin.ca> and the project contributors
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package resources
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/purpleidea/mgmt/engine"
+	"github.com/purpleidea/mgmt/engine/traits"
+	"github.com/purpleidea/mgmt/util"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/api/types/filters"
+	"github.com/docker/docker/client"
+	"github.com/docker/go-connections/nat"
+)
+
+const (
+	// ContainerRunning is the running container state.
+	ContainerRunning = "running"
+	// ContainerStopped is the stopped container state.
+	ContainerStopped = "stopped"
+	// ContainerRemoved is the removed container state.
+	ContainerRemoved = "removed"
+
+	// dockerClientVersion is the api version used by the client. See:
+	// https://docs.docker.com/develop/sdk/#api-version-matrix for more info.
+	dockerClientVersion = "v1.37"
```

Comment:
So it bugged me that we'd have some constant in here which would potentially rot over time, so I did some source diving, and: api.DefaultVersion exists! woo. But even better, it seems if we use the empty string, then that is taken as the default. Woo, woo!

So I think you should move this to a struct arg that defaults to empty string. Unless it's never something you want to specify in which case just empty string. Also there's this: https://github.com/moby/moby/blob/e6244aa598e9894ead8445ca66329d26eb016e64/client/client.go#L249

What do you think? I'm leaning on struct option + warning to specify a version if desired.

Reply:
My only concern is that the way certain methods work, or if they're even available at all, may change depending on the API version, leading to undefined behaviour. I think `api.DefaultVersion` is probably the safest bet, and fits with your "do it with git master" philosophy.

Reply:
:(
```
18:43:23 main.go:186: main: error validating the new graph: error searching for image: Error response from daemon: client version 1.40 is too new. Maximum supported API version is 1.39
```